cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/SingleThreadDatasetCache.java

```diff
@@ -0,0 +1,206 @@
+  private final Map<DatasetCacheKey, TransactionAware> inProgress = new HashMap<>();
+  private final Map<DatasetCacheKey, Dataset> staticDatasets = new HashMap<>();
+  private TransactionContext txContext = null;
+
+  /**
+   * See {@link DynamicDatasetCache}.
+   *
+   * @param staticDatasets  if non-null, a map from dataset name to runtime arguments. These datasets will be
+   *                        instantiated immediately, and they will participate in every transaction started
+   *                        through {@link #newTransactionContext()}.
+   */
+  public SingleThreadDatasetCache(final SystemDatasetInstantiator instantiator,
+                                  final TransactionSystemClient txClient,
+                                  final Id.Namespace namespace,
+                                  Map<String, String> runtimeArguments,
+                                  @Nullable final MetricsContext metricsContext,
+                                  @Nullable Map<String, Map<String, String>> staticDatasets) {
+    super(instantiator, txClient, namespace, runtimeArguments, metricsContext);
+    this.datasetLoader = new CacheLoader<DatasetCacheKey, Dataset>() {
+      @Override
+      @ParametersAreNonnullByDefault
+      public Dataset load(DatasetCacheKey key) throws Exception {
+        Dataset dataset = instantiator.getDataset(
+          Id.DatasetInstance.from(namespace, key.getName()), key.getArguments());
+        if (dataset instanceof MeteredDataset && metricsContext != null) {
+          ((MeteredDataset) dataset).setMetricsCollector(
+            metricsContext.childContext(Constants.Metrics.Tag.DATASET, key.getName()));
+        }
+        return dataset;
+      }
+    };
+    this.datasetCache = CacheBuilder.newBuilder().removalListener(
+      new RemovalListener<DatasetCacheKey, Dataset>() {
+        @Override
+        @ParametersAreNonnullByDefault
+        public void onRemoval(RemovalNotification<DatasetCacheKey, Dataset> notification) {
+          if (notification.getValue() != null) {
+            try {
+              notification.getValue().close();
+            } catch (Throwable e) {
+              LOG.warn("Error closing dataset '{}'", notification.getKey());
+            }
+          }
+        }
+      })
+      .build(datasetLoader);
+    if (staticDatasets != null) {
+      for (Map.Entry<String, Map<String, String>> entry : staticDatasets.entrySet()) {
+        this.staticDatasets.put(new DatasetCacheKey(entry.getKey(), entry.getValue()),
+                                getDataset(entry.getKey(), entry.getValue()));
+      }
+    }
+  }
+
+  @Override
+  public synchronized <T extends Dataset> T getDataset(DatasetCacheKey key, boolean bypass)
+    throws DatasetInstantiationException {
+
+    Dataset dataset;
+    try {
+      if (bypass) {
+        dataset = datasetLoader.load(key);
+      } else {
+        try {
+          dataset = datasetCache.get(key);
```

Comment:
Should we always check the `staticDatasets` map first and return the one from there if found?